### ISSUE TITLE
pointycastle upgrade and asn1lib migration

### ISSUE THREAD

**Issue by awhitford** [Ephenodrom/Dart-Basic-Utils#31]
It caught my eye that `pubspec.yaml` still referenced old dependencies:
```
-  pointycastle: ^1.0.2
-  asn1lib: ^0.6.5
```
Even more fascinating is that `pointycastle: ^2.0.0` _includes_ an `asn1` implementation, so the `asn1lib` dependency could be dropped/replaced -- but I know you know this @Ephenodrom.

For fun, I tried to _upgrade_ this package, but I ran into 3 roadblocks that I cannot seem to overcome.  It appears that _pointycastle's asn1_ implementation is not quite ready for prime time.

However, I thought it would be worth sharing my discoveries to date to aid the eventual migration:

1. Calls to `ASN1ObjectIdentifier.registerFrequentNames();` may be removed; there does not seem to be an equivalent.  (And honestly, this strikes me as an odd API design for `asn1lib`, but I am not an expert on ASN1.) 👍 
2. Most old calls to `.encodedBytes` need to be replaced with calls to `.encode()`.  (It looks like the old ASN1lib would trigger an encoding if `encodedBytes` was called before `encode()`, but the new implementation simply returns `null`. A getter should not maybe do something expensive, so I'm OK with `encodedBytes` being a simple getter, but I would expect it to _assert_ if it is illegal to call before `encode()`.)
3. `ASN1BitString` constructor needs to add parameter name like `stringValues`.  (Honestly, I would rethink this to be more backwards compatible with the old `asn1lib`.  You can have alternate constructors or factory methods to handle different scenarios.)
4. `ASN1OctetString` constructor needs to add parameter name like `octets`.  (Same issue as `ASN1BitString`.)
5. Calls to `.contentBytes()`. need to be replaced with `.valueBytes`.
6. Calls to `.valueAsBigInteger` need to replaced with `.integer`.
7. Calls to `.identifier` need to be replaced with `.objectIdentifierAsString`.
8. For `ASN1UtcTime`, calls to `.dateTimeValue` need to be replaced with `.time`.
9. Calls to `.valueBytes()` need to be replaced with `.valueBytes`.
10. Curve Name translations can now be radically simplified from:
```dart
var curveNameOi = s.elements.elementAt(0) as ASN1ObjectIdentifier;
var curveName;
ASN1ObjectIdentifier.DN.keys.forEach((element) {
  if (ASN1ObjectIdentifier.DN[element] == curveNameOi.identifier) {
    curveName = element;
  }
});
```
to:
```dart
var curveNameOi = s.elements.elementAt(0) as ASN1ObjectIdentifier;
final curveName = curveNameOi.readableName;
``` 
🎉 

After all these changes, I could not get past 9 unit test failures.  Of these 9 failures, there are basically 3 outstanding issues, all of which seem to be related to _pointycastle's ASN1 implementation_:
1. This unit test failure:
```
00:13 +54 -1: test/x509_utils_test.dart: Test generateRsaCsrPem [E]                                                                        
  NoSuchMethodError: The getter 'iterator' was called on null.
  Receiver: null
  Tried calling: iterator
  dart:core                                                        List.addAll
  package:pointycastle/asn1/primitives/asn1_bit_string.dart 82:11  ASN1BitString.encode
  package:pointycastle/asn1/primitives/asn1_sequence.dart 69:16    ASN1Sequence._childLength.<fn>
  dart:core                                                        List.forEach
  package:pointycastle/asn1/primitives/asn1_sequence.dart 68:14    ASN1Sequence._childLength
  package:pointycastle/asn1/primitives/asn1_sequence.dart 51:25    ASN1Sequence.encode
  package:pointycastle/asn1/primitives/asn1_sequence.dart 69:16    ASN1Sequence._childLength.<fn>
  dart:core                                                        List.forEach
  package:pointycastle/asn1/primitives/asn1_sequence.dart 68:14    ASN1Sequence._childLength
  package:pointycastle/asn1/primitives/asn1_sequence.dart 51:25    ASN1Sequence.encode
  package:basic_utils/src/X509Utils.dart 125:54                    X509Utils.generateRsaCsrPem
  test/x509_utils_test.dart 279:25                                 main.<fn>
```
The ultimate code is simply:
```dart
    var blockDN = ASN1Sequence();
    blockDN.add(ASN1Integer(BigInt.from(0)));
    blockDN.add(encodedDN);
    blockDN.add(_makePublicKeyBlock(publicKey));
    blockDN.add(ASN1Null(tag: 0xA0)); // let's call this WTF
...
    blockDN.encode()
```
The `blockDN` setup is no different -- the old code called `blockDN.encodedBytes` instead.

2. Incorrect Public Key Length -- `pubKeyLength` is 2168 instead of 2048, for example.  The value seems to be consistently +120 (a 4096 became 4216), but I cannot figure out what is wrong.  This appears to be an issue with _pointycastle's ASN1 implementation_.

3. I can't seem to get this to work:
```dart
  static RSAPublicKey rsaPublicKeyFromDERBytes(Uint8List bytes) {
    var asn1Parser = ASN1Parser(bytes);
    var topLevelSeq = asn1Parser.nextObject() as ASN1Sequence;
    var publicKeyBitString = topLevelSeq.elements[1];

    var publicKeyAsn = ASN1Parser(publicKeyBitString.valueBytes);
    ASN1Sequence publicKeySeq = publicKeyAsn.nextObject();
    var modulus = publicKeySeq.elements[0] as ASN1Integer;
    var exponent = publicKeySeq.elements[1] as ASN1Integer;

    var rsaPublicKey = RSAPublicKey(modulus.integer, exponent.integer);

    return rsaPublicKey;
  }
```
It says:
```
  UnsupportedASN1TagException: Tag 0 is not supported yet
  package:pointycastle/asn1/asn1_parser.dart 149:9  ASN1Parser._createPrimitive
  package:pointycastle/asn1/asn1_parser.dart 78:13  ASN1Parser.nextObject
  package:basic_utils/src/CryptoUtils.dart 446:46   CryptoUtils.rsaPublicKeyFromDERBytes
  test/crypto_utils_test.dart 170:30                main.<fn>
```

I hope this helps the eventual upgrade/migration.

**Comment by Ephenodrom** [comment]
@awhitford The migration to the new Pointycastle version is already done on my local PC. I am using the latest RC2 version available on pub.dev. I am just waiting for the final release of 2.0.0.

It took me only 30 minutes to rewrite the code. The new asn1lib from pointycastle has a bit easier structure, that's why i had to change some code.

All unit tests are working fine.

So stay tuned for the update.

**Comment by awhitford** [comment]
@Ephenodrom That sounds great.  I'm patient.  😄

**Comment by Ephenodrom** [comment]
@awhitford  Release 2.7.0-rc.1 is out now. Please try it out. Feel free to report any problems. I will create another release candidate if the "NPE" bug in the PointyCastle ASN1 Lib is fixed by me.

Regards

**Comment by Ephenodrom** [comment]
This is now fixed with the version ^2.7.0-rc.5, due to it includes Pointycastle 2.0.1.
